python/ray/autoscaler/_private/spark/node_provider.py

```diff
@@ -0,0 +1,215 @@
+
+    def is_running(self, node_id):
+        with self.lock:
+            return (
+                node_id in self._nodes
+                and self._nodes[node_id]["tags"][TAG_RAY_NODE_STATUS]
+                == STATUS_UP_TO_DATE
+            )
+
+    def is_terminated(self, node_id):
+        with self.lock:
+            return node_id not in self._nodes
+
+    def node_tags(self, node_id):
+        with self.lock:
+            return self._nodes[node_id]["tags"]
+
+    def _get_ip(self, node_id: str) -> Optional[str]:
+        return node_id
+
+    def external_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def internal_ip(self, node_id):
+        return self._get_ip(node_id)
+
+    def set_node_tags(self, node_id, tags):
+        assert node_id in self._nodes
+        self._nodes[node_id]["tags"].update(tags)
+
+    def create_node(
+        self, node_config: Dict[str, Any], tags: Dict[str, str], count: int
+    ) -> Optional[Dict[str, Any]]:
+        raise AssertionError("This method should not be called.")
+
+    def _gen_spark_job_group_id(self, node_id):
+        return (
+            f"ray-cluster-{self.ray_head_port}-{self.cluster_unique_id}"
+            f"-worker-node-{node_id}"
+        )
+
+    def create_node_with_resources_and_labels(
+        self, node_config, tags, count, resources, labels
+    ):
+        from ray.util.spark.cluster_init import _append_resources_config
+
+        # Note: even if count > 1 , we should only create 1 node in this call
+        #  otherwise some issue happens in my test.
+        with self.lock:
+            resources = resources.copy()
+            node_type = tags[TAG_RAY_USER_NODE_TYPE]
+            # NOTE:
+            #  "NODE_ID_AS_RESOURCE" value must be an integer,
+            #  but `node_id` used by autoscaler must be a string.
+            node_id = str(self.get_next_node_id())
+            resources["NODE_ID_AS_RESOURCE"] = int(node_id)
+
+            conf = self.provider_config
+
+            num_cpus_per_node = resources.pop("CPU")
+            num_gpus_per_node = resources.pop("GPU")
+            heap_memory_per_node = resources.pop("memory")
+            object_store_memory_per_node = resources.pop("object_store_memory")
+
+            conf["worker_node_options"] = _append_resources_config(
+                conf["worker_node_options"], resources
+            )
+            response = requests.post(
+                url=self.server_url + "/create_node",
+                json={
+                    "spark_job_group_id": self._gen_spark_job_group_id(node_id),
+                    "spark_job_group_desc": (
+                        "This job group is for spark job which runs the Ray "
+                        f"cluster worker node {node_id} connecting to ray "
+                        f"head node {self.ray_head_ip}:{self.ray_head_port}"
+                    ),
+                    "using_stage_scheduling": conf["using_stage_scheduling"],
```

Comment:
Why do we need to pass these provider configs to server. When we create the server, these can be already passed in since these are not per node configs?

Reply:
> Why do we need to pass these provider configs to server ?

For current code, we don't need to pass these to server, but in future I might want to merge multiple server into one, and passing all configs per requests makes this change in future easier

also, keeping these configs in autoscaling config file makes us easier to debug, do we have to remove them for some reason ?